# Customer Experience & E-Commerce Performance Investigation

## Phase 1 — Data Understanding

### Objective
Understand the structure, grain, relationships, keys, data types, and initial data-quality characteristics of the Olist Brazilian E-Commerce dataset before cleaning or analysis.

In [1]:
import pandas as pd
import numpy as np

In [3]:
customers = pd.read_csv("../data/raw/olist_customers_dataset.csv")
orders = pd.read_csv("../data/raw/olist_orders_dataset.csv")
order_items = pd.read_csv("../data/raw/olist_order_items_dataset.csv")
payments = pd.read_csv("../data/raw/olist_order_payments_dataset.csv")
reviews = pd.read_csv("../data/raw/olist_order_reviews_dataset.csv")
products = pd.read_csv("../data/raw/olist_products_dataset.csv")
sellers = pd.read_csv("../data/raw/olist_sellers_dataset.csv")
geolocation = pd.read_csv("../data/raw/olist_geolocation_dataset.csv")
category_translation = pd.read_csv("../data/raw/product_category_name_translation.csv")

In [4]:
tables = {
    "customers": customers,
    "orders": orders,
    "order_items": order_items,
    "payments": payments,
    "reviews": reviews,
    "products": products,
    "sellers": sellers,
    "geolocation": geolocation,
    "category_translation": category_translation
}

for name, df in tables.items():
    print(f"{name}: {df.shape[0]:,} rows × {df.shape[1]} columns")

customers: 99,441 rows × 5 columns
orders: 99,441 rows × 8 columns
order_items: 112,650 rows × 7 columns
payments: 103,886 rows × 5 columns
reviews: 99,224 rows × 7 columns
products: 32,951 rows × 9 columns
sellers: 3,095 rows × 4 columns
geolocation: 1,000,163 rows × 5 columns
category_translation: 71 rows × 2 columns


In [5]:
for name, df in tables.items():
    print(f"\n{'=' * 60}")
    print(name.upper())
    print(f"{'=' * 60}")
    print(list(df.columns))


CUSTOMERS
['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']

ORDERS
['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']

ORDER_ITEMS
['order_id', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value']

PAYMENTS
['order_id', 'payment_sequential', 'payment_type', 'payment_installments', 'payment_value']

REVIEWS
['review_id', 'order_id', 'review_score', 'review_comment_title', 'review_comment_message', 'review_creation_date', 'review_answer_timestamp']

PRODUCTS
['product_id', 'product_category_name', 'product_name_lenght', 'product_description_lenght', 'product_photos_qty', 'product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm']

SELLERS
['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state']

GEOLOCATION
['geolocation_zi

In [6]:
customers.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [7]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [8]:
customers.nunique()

customer_id                 99441
customer_unique_id          96096
customer_zip_code_prefix    14994
customer_city                4119
customer_state                 27
dtype: int64

In [9]:
customers["customer_unique_id"].value_counts().head(10)

customer_unique_id
8d50f5eadf50201ccdcedfb9e2ac8455    17
3e43e6105506432c953e165fb2acf44c     9
1b6c7548a2a1f9037c1fd3ddfed95f33     7
ca77025e7201e3b30c44b472ff346268     7
6469f99c1f9dfae7733b25662e7f1782     7
63cfc61cee11cbe306bff5857d00bfe4     6
47c1a3033b8b77b3ab6e109eb4d5fdf3     6
12f5d6e1cbf93dafd9dcc19095df0b3d     6
de34b16117594161a6a89c50b289d35a     6
dc813062e0fc23409cd255f7f53c7074     6
Name: count, dtype: int64

In [10]:
orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [11]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [12]:
orders.nunique()

order_id                         99441
customer_id                      99441
order_status                         8
order_purchase_timestamp         98875
order_approved_at                90733
order_delivered_carrier_date     81018
order_delivered_customer_date    95664
order_estimated_delivery_date      459
dtype: int64

In [13]:
orders["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

### Orders — Initial Observations

- Grain: one row represents one order.
- Primary key: `order_id`.
- Foreign key: `customer_id`, linking orders to the customers table.
- The table contains 99,441 orders and 8 columns.
- `order_status` contains 8 distinct statuses.
- Most orders are delivered, while a smaller number are shipped, canceled, unavailable, invoiced, processing, created, or approved.
- Several delivery-related fields contain missing values; these will be investigated during data cleaning rather than changed during data understanding.
- Date/time columns are currently stored as object/string types and will be converted during data cleaning.

In [14]:
order_items.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


In [15]:
order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [16]:
order_items.nunique()

order_id               98666
order_item_id             21
product_id             32951
seller_id               3095
shipping_limit_date    93318
price                   5968
freight_value           6999
dtype: int64

In [17]:
order_items.duplicated(
    subset=["order_id", "order_item_id"]
).sum()

0

In [18]:
order_items.groupby("order_id")["order_item_id"].count().value_counts().sort_index()

order_item_id
1     88863
2      7516
3      1322
4       505
5       204
6       198
7        22
8         8
9         3
10        8
11        4
12        5
13        1
14        2
15        2
20        2
21        1
Name: count, dtype: int64

### Order Items — Initial Observations

- Grain: one row represents one item within an order.
- `order_id` is not unique because an order can contain multiple items.
- `order_item_id` identifies the item's sequence within an order and is not unique by itself.
- The combination of `order_id` and `order_item_id` uniquely identifies each order-item record.
- `order_id` links the order-item records back to the `orders` table.
- `product_id` links an order item to the `products` table.
- `seller_id` links an order item to the `sellers` table.
- Price and freight information are stored at the order-item level.
- The dataset contains 112,650 order-item records across 98,666 distinct orders.

In [19]:
payments.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


In [20]:
payments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


In [21]:
payments.nunique()

order_id                99440
payment_sequential         29
payment_type                5
payment_installments       24
payment_value           29077
dtype: int64

In [22]:
payments.groupby("order_id")["payment_sequential"].count().value_counts().sort_index()

payment_sequential
1     96479
2      2382
3       301
4       108
5        52
6        36
7        28
8        11
9         9
10        5
11        8
12        8
13        3
14        2
15        2
19        2
21        1
22        1
26        1
29        1
Name: count, dtype: int64

In [23]:
payments.duplicated(
    subset=["order_id", "payment_sequential"]
).sum()

0

## 💳 Payments Table — Phase 1 Understanding

- **Rows:** 103,886
- **Columns:** 5
- **Grain:** One row represents one payment record associated with an order.
- **`order_id`:** Identifies the order, but is not unique because an order can have multiple payment records.
- **`payment_sequential`:** Represents the sequence number of a payment within an order.
- **`payment_installments`:** Represents the number of installments for a payment; it is not the number of payment records.
- **`payment_type`:** Represents the payment method.
- **`payment_value`:** Represents the value of the payment record.
- **Composite key:** `order_id` + `payment_sequential`.
- **Key validation:** The combination of `order_id` and `payment_sequential` has 0 duplicate records, confirming that the combination uniquely identifies a payment record.

### Key takeaway
The Payments table has a **one-to-many relationship with Orders**: one order can have multiple payment records.

In [24]:
reviews.head()

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


In [25]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


In [26]:
len(reviews)

99224

In [27]:
reviews["review_id"].nunique()

98410

In [28]:
reviews["order_id"].nunique()

98673

In [29]:
reviews.duplicated(subset=["review_id"]).sum()

814

In [30]:
reviews.duplicated(subset=["review_id", "order_id"]).sum()

0

In [31]:
reviews.groupby("order_id")["review_id"].count().value_counts().sort_index()

review_id
1    98126
2      543
3        4
Name: count, dtype: int64

## ⭐ Reviews Table — Phase 1 Understanding

- **Rows:** 99,224
- **Columns:** 7
- **Grain:** One row represents one review record associated with an order.
- **`order_id`:** Connects Reviews to the Orders table. It is not unique because some orders have multiple review records.
- **`review_id`:** Identifies a review, but it is not unique in this dataset; 814 duplicate `review_id` records were observed.
- **Composite key:** `review_id` + `order_id`.
- **Key validation:** The combination of `review_id` and `order_id` has 0 duplicate records, so the combination uniquely identifies the review record in this dataset.
- **`review_score`:** Main customer satisfaction measure, ranging from the customer's rating of the order.
- **`review_comment_title` and `review_comment_message`:** Customer-written feedback used later for sentiment, complaint taxonomy, and theme analysis.
- **Missing review text:** Many reviews do not contain a title or message. These missing values will be investigated and handled during the data-cleaning/NLP phases rather than removed automatically.

### Key takeaway
The Reviews table provides the primary customer-experience feedback used to investigate what factors are associated with positive and negative customer satisfaction.

In [32]:
products.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


In [33]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [34]:
len(products)

32951

In [35]:
products["product_id"].nunique()

32951

In [36]:
products.duplicated(subset=["product_id"]).sum()

0

In [37]:
order_items["product_id"].nunique()

32951

## 📦 Products Table — Phase 1 Understanding

- **Rows:** 32,951
- **Columns:** 9
- **Grain:** One row represents one product.
- **`product_id`:** Uniquely identifies a product.
- **Key validation:** `product_id` has 32,951 unique values and 0 duplicates, confirming that it uniquely identifies each product in this table.
- **`product_category_name`:** Represents the product's category.
- **Product attributes:** Include name length, description length, number of photos, weight, length, height, and width.
- **Missing values:** Product category and related descriptive fields contain missing values, while a small number of products have missing physical measurements.
- **Relationship:** `product_id` connects Products to the Order Items table.
- **Relationship type:** One product can appear in multiple order-item records.

### Key takeaway
The Products table provides product-level attributes that will later be used to investigate product categories, product characteristics, freight/delivery behavior, and customer satisfaction.

In [38]:
sellers.head()

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


In [39]:
sellers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


In [40]:
len(sellers)

3095

In [41]:
sellers["seller_id"].nunique()

3095

In [42]:
sellers.duplicated(subset=["seller_id"]).sum()

0

## 🏪 Sellers Table — Phase 1 Understanding

- **Rows:** 3,095
- **Columns:** 4
- **Grain:** One row represents one seller.
- **`seller_id`:** Uniquely identifies a seller.
- **Key validation:** `seller_id` has 3,095 unique values and 0 duplicates.
- **`seller_zip_code_prefix`:** Represents the seller's geographic location prefix.
- **`seller_city` and `seller_state`:** Represent the seller's location.
- **Missing values:** No missing values were observed in the Sellers table.
- **Relationship:** `seller_id` connects Sellers to the Order Items table.
- **Relationship type:** One seller can appear in multiple order-item records.

### Key takeaway
The Sellers table provides seller-level identification and geographic information that will later be used to evaluate seller performance, delivery behavior, revenue, and customer satisfaction.

In [43]:
geolocation.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


In [44]:
geolocation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB


In [45]:
geolocation["geolocation_zip_code_prefix"].nunique()

19015

In [46]:
geolocation.duplicated().sum()

261831

In [47]:
geolocation.duplicated(
    subset=["geolocation_zip_code_prefix"]
).sum()

981148

## 🌎 Geolocation Table — Phase 1 Understanding

- **Rows:** 1,000,163
- **Columns:** 5
- **Grain:** One row represents a geolocation record containing a ZIP-code prefix and associated geographic information.
- **`geolocation_zip_code_prefix`:** Identifies a ZIP-code prefix but is not unique; 19,015 unique prefixes exist across more than 1 million records.
- **Geographic attributes:** Include latitude, longitude, city, and state.
- **Duplicate records:** 261,831 exact duplicate rows were observed.
- **ZIP-prefix duplicates:** 981,148 rows contain ZIP prefixes that appear in earlier records.
- **Primary key:** No single unique identifier is present in the table based on the columns inspected so far.
- **Relationship:** The ZIP-code prefix can potentially be used to connect geographic information with customer and seller location data.
- **Data-quality consideration:** Multiple records can exist for the same ZIP prefix, including records with different geographic coordinates. This needs to be handled carefully during data preparation.

### Key takeaway
The Geolocation table contains multiple geographic observations for the same ZIP-code prefixes rather than one unique record per ZIP code. It will later support geographic analysis of customers and sellers.

In [48]:
category_translation.head()

,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


In [49]:
category_translation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


In [50]:
len(category_translation)

71

In [51]:
category_translation.duplicated().sum()

0

In [52]:
category_translation["product_category_name"].nunique()

71

In [53]:
category_translation["product_category_name_english"].nunique()

71

## 🌐 Category Translation Table — Phase 1 Understanding

- **Rows:** 71
- **Columns:** 2
- **Grain:** One row represents one product-category translation mapping.
- **`product_category_name`:** Original Portuguese product category name.
- **`product_category_name_english`:** English translation of the product category.
- **Missing values:** No missing values were observed.
- **Duplicate complete rows:** 0 duplicate rows were observed.
- **Unique categories:** Both category columns contain 71 unique values.
- **Lookup key:** `product_category_name` can uniquely identify each translation mapping.
- **Purpose:** Provides an English mapping for the original product categories in the Products table.
- **Relationship:** `product_category_name` connects the translation table to the Products table.

### Key takeaway
The Category Translation table is a small lookup table that maps the original Portuguese product categories to English categories, making category analysis and dashboard presentation easier to interpret.

In [54]:
customers["customer_id"].nunique()

99441

In [55]:
orders["customer_id"].nunique()

99441

In [56]:
orders["customer_id"].isin(customers["customer_id"]).all()

True

### 🔗 Customers → Orders Relationship

- `customer_id` is unique in the Customers table.
- Orders contains 99,441 unique `customer_id` values.
- Every `customer_id` in Orders exists in Customers, confirming referential integrity.
- Each `customer_id` appears once in Orders in this dataset.
- However, `customer_id` represents an order-associated customer record, while `customer_unique_id` represents the underlying customer identity.
- Multiple customer records can share the same `customer_unique_id`.
- Therefore, `customer_unique_id` is important for identifying repeat customers and analyzing repeat-purchase behavior.

In [61]:
customers["customer_unique_id"].nunique()

96096

In [62]:
orders["order_id"].nunique()

99441

In [63]:
order_items["order_id"].nunique()

98666

In [64]:
order_items["order_id"].isin(orders["order_id"]).all()

True

### 🔗 Orders → Order Items Relationship

- `order_id` uniquely identifies an order in the Orders table.
- Order Items contains 98,666 unique order IDs compared with 99,441 unique orders in Orders.
- Every `order_id` appearing in Order Items exists in Orders, confirming referential integrity from Order Items to Orders.
- Some orders do not have corresponding Order Items records; this difference will be investigated later rather than assumed to be an error.
- An order can contain multiple order-item records.
- Therefore, the relationship between Orders and Order Items is 1-to-many.

### Key takeaway
Orders provides order-level information, while Order Items provides item-level information such as product, seller, price, and freight value.

In [65]:
orders["order_id"].nunique()

99441

In [66]:
payments["order_id"].nunique()

99440

In [67]:
payments["order_id"].isin(orders["order_id"]).all()

True

In [68]:
payments.groupby("order_id")["payment_sequential"].count().value_counts().sort_index()

payment_sequential
1     96479
2      2382
3       301
4       108
5        52
6        36
7        28
8        11
9         9
10        5
11        8
12        8
13        3
14        2
15        2
19        2
21        1
22        1
26        1
29        1
Name: count, dtype: int64

### 🔗 Orders → Payments Relationship

- `order_id` uniquely identifies an order in the Orders table.
- Payments contains 99,440 unique order IDs compared with 99,441 unique orders in Orders.
- Every `order_id` appearing in Payments exists in Orders, confirming referential integrity.
- Some orders have multiple payment records, with the number of records per order ranging from 1 to 29.
- `payment_sequential` identifies the sequence of payment records within an order.
- `payment_installments` represents the number of installments for a payment and is separate from the number of payment records.
- Therefore, the relationship between Orders and Payments is 1-to-many.

### Key takeaway
The Payments table provides payment-level information, while Orders provides order-level information. Multiple payment records can be associated with a single order.

### ⚠️ Data Observation to Investigate Later
- Orders contains **99,441 unique orders**, while Payments contains **99,440 unique order IDs**.
- This means one order does not currently appear in the Payments table.
- We will investigate this discrepancy during the data-quality/cleaning phase rather than assuming its cause at this stage.

In [69]:
reviews["order_id"].isin(orders["order_id"]).all()

True

In [70]:
reviews["order_id"].nunique()

98673

In [71]:
orders["order_id"].isin(reviews["order_id"]).sum()

98673

### 🔗 Orders → Reviews Relationship

- Every `order_id` appearing in Reviews exists in Orders, confirming referential integrity.
- Reviews contains **98,673 unique order IDs** across **99,224 review records**.
- Some orders therefore have multiple review records.
- Based on the review distribution:
  - 98,126 orders have 1 review.
  - 543 orders have 2 reviews.
  - 4 orders have 3 reviews.
- Orders contains 99,441 unique orders, of which 98,673 have at least one review.
- Therefore, **768 orders do not have a review record**.
- The relationship from Orders to Reviews is **1-to-many with optional reviews**: one order can have zero, one, or multiple review records.

### Key takeaway
The Reviews table provides customer satisfaction and written feedback for orders, but not every order has a review. This distinction will be important when calculating customer satisfaction metrics and analyzing review-based results.

In [72]:
order_items["product_id"].isin(products["product_id"]).all()

True

In [73]:
order_items["product_id"].nunique()

32951

In [74]:
order_items["seller_id"].isin(sellers["seller_id"]).all()

True

In [76]:
order_items["seller_id"].nunique()

3095

### 🔗 Order Items → Sellers Relationship

- Every `seller_id` appearing in Order Items exists in the Sellers table, confirming referential integrity.
- Order Items contains **3,095 unique seller IDs**, matching the **3,095 sellers** in the Sellers table.
- A seller can appear in multiple order-item records because sellers can sell products across multiple orders.
- Therefore, the relationship between Sellers and Order Items is **1-to-many**.

### Key takeaway
The Sellers table provides seller-level identification and location information, while Order Items records the products sold by each seller along with price and freight information.

In [77]:
products["product_category_name"].isin(
    category_translation["product_category_name"]
).value_counts()

product_category_name
True     32328
False      623
Name: count, dtype: int64

In [78]:
products["product_category_name"].nunique()

73

In [79]:
category_translation["product_category_name"].nunique()

71

In [80]:
products.loc[
    products["product_category_name"].notna()
    & ~products["product_category_name"].isin(
        category_translation["product_category_name"]
    ),
    "product_category_name"
].value_counts()

product_category_name
portateis_cozinha_e_preparadores_de_alimentos    10
pc_gamer                                          3
Name: count, dtype: int64

## 🌐 Category Translation Table — Phase 1 Understanding

- **Rows:** 71
- **Columns:** 2
- **Grain:** One row represents one product-category translation mapping.
- **`product_category_name`:** Original Portuguese product category name.
- **`product_category_name_english`:** English translation of the product category.
- **Missing values:** No missing values were observed in the translation table.
- **Duplicate complete rows:** 0 duplicate rows were observed.
- **Unique categories:** Both category columns contain 71 unique values.
- **Lookup key:** `product_category_name` uniquely identifies each translation mapping.
- **Purpose:** Provides an English mapping for the original Portuguese product categories in the Products table.
- **Relationship:** `product_category_name` connects the translation table to the Products table.

### ⚠️ Data Observation

- Products contains **73 non-null product categories**, while the translation table contains **71 categories**.
- Two product categories do not have English translations:
  - `portateis_cozinha_e_preparadores_de_alimentos` — 10 product records
  - `pc_gamer` — 3 product records
- Products also contains **610 records with missing `product_category_name`**.
- These missing categories and untranslated categories will be handled during the data-cleaning phase rather than being silently removed.

### Key takeaway
The Category Translation table acts as a lookup table for converting Portuguese product categories to English. Most categories have translations, but two categories require special handling during data preparation.

In [81]:
customers["customer_zip_code_prefix"].nunique()

14994

In [82]:
customers["customer_zip_code_prefix"].isin(
    geolocation["geolocation_zip_code_prefix"]
).all()

False

In [83]:
customers["customer_zip_code_prefix"].isin(
    geolocation["geolocation_zip_code_prefix"]
).value_counts()

customer_zip_code_prefix
True     99163
False      278
Name: count, dtype: int64

In [84]:
sellers["seller_zip_code_prefix"].nunique()

2246

In [85]:
sellers["seller_zip_code_prefix"].isin(
    geolocation["geolocation_zip_code_prefix"]
).all()

False

In [86]:
sellers["seller_zip_code_prefix"].isin(
    geolocation["geolocation_zip_code_prefix"]
).value_counts()

seller_zip_code_prefix
True     3088
False       7
Name: count, dtype: int64

### 🔗 Customers/Sellers → Geolocation Relationship

#### Customers → Geolocation

- Customers contains **14,994 unique customer ZIP-code prefixes**.
- 99,163 customer records have a ZIP prefix found in Geolocation.
- 278 customer records have a ZIP prefix not found in Geolocation.
- Therefore, not every customer ZIP prefix has a corresponding record in the Geolocation table.

#### Sellers → Geolocation

- Sellers contains **2,246 unique seller ZIP-code prefixes**.
- 3,088 seller records have a ZIP prefix found in Geolocation.
- 7 seller records have a ZIP prefix not found in Geolocation.
- Therefore, not every seller ZIP prefix has a corresponding record in the Geolocation table.

#### Relationship characteristics

- Customers and Sellers connect to Geolocation through their ZIP-code prefix.
- `geolocation_zip_code_prefix` is not unique in the Geolocation table.
- Multiple geolocation records can correspond to the same ZIP-code prefix.
- Therefore, Geolocation is treated as a supporting geographic lookup/source rather than a table with a simple one-to-one relationship.
- Unmatched ZIP prefixes will be investigated during the data-quality/cleaning phase rather than assumed to be invalid.

### Key takeaway
The Geolocation table can enrich customer and seller records with geographic information, but its repeated ZIP-code prefixes and unmatched records require careful handling during data preparation.

In [87]:
orders["order_purchase_timestamp"].min(), orders["order_purchase_timestamp"].max()

('2016-09-04 21:15:19', '2018-10-17 17:30:18')

In [ ]:
orders["order_estimated_delivery_date"].min(), orders["order_estimated_delivery_date"].max()

In [89]:
pd.to_datetime(
    orders["order_delivered_customer_date"],
    errors="coerce"
).min(), pd.to_datetime(
    orders["order_delivered_customer_date"],
    errors="coerce"
).max()

(Timestamp('2016-10-11 13:46:32'), Timestamp('2018-10-17 13:22:46'))

In [90]:
pd.to_datetime(
    orders["order_estimated_delivery_date"],
    errors="coerce"
).min(), pd.to_datetime(
    orders["order_estimated_delivery_date"],
    errors="coerce"
).max()

(Timestamp('2016-09-30 00:00:00'), Timestamp('2018-11-12 00:00:00'))

In [91]:
for name, df in tables.items():
    print(name, ":", df.duplicated().sum())

customers : 0
orders : 0
order_items : 0
payments : 0
reviews : 0
products : 0
sellers : 0
geolocation : 261831
category_translation : 0


# 🔗 Phase 1 — Overall Data Model

## Core Relationships

### Customers → Orders
- `customers.customer_id` → `orders.customer_id`
- Every order customer ID exists in Customers.
- `customer_id` is unique in both tables in this dataset.
- `customer_unique_id` represents the underlying customer identity and is important for repeat-purchase analysis.

### Orders → Order Items
- `orders.order_id` → `order_items.order_id`
- One order can contain multiple order-item records.
- 775 orders are not represented in Order Items and will be investigated during data cleaning.

### Orders → Payments
- `orders.order_id` → `payments.order_id`
- One order can have multiple payment records.
- One order is not represented in Payments and will be investigated later.

### Orders → Reviews
- `orders.order_id` → `reviews.order_id`
- One order can have zero, one, or multiple review records.
- 768 orders have no review record.

### Order Items → Products
- `order_items.product_id` → `products.product_id`
- Every product ID in Order Items exists in Products.
- One product can appear in multiple order-item records.

### Order Items → Sellers
- `order_items.seller_id` → `sellers.seller_id`
- Every seller ID in Order Items exists in Sellers.
- One seller can appear in multiple order-item records.

### Products → Category Translation
- `products.product_category_name` → `category_translation.product_category_name`
- The translation table provides English category names.
- Two product categories do not have translations.

### Customers/Sellers → Geolocation
- `customer_zip_code_prefix` and `seller_zip_code_prefix` can be matched to `geolocation_zip_code_prefix`.
- Geolocation contains multiple records per ZIP prefix.
- Some customer and seller ZIP prefixes have no matching geolocation record.
- Geolocation is therefore treated as a supporting geographic data source rather than a simple one-to-one lookup table.

# 📚 Phase 1 — Data Dictionary

| Table | Grain | Primary/Unique Identifier | Main Purpose |
|---|---|---|---|
| Customers | One row per customer record | `customer_id` | Customer identity and location |
| Orders | One row per order | `order_id` | Order lifecycle and dates |
| Order Items | One row per item within an order | `order_id` + `order_item_id` | Products, sellers, price and freight |
| Payments | One row per payment record | `order_id` + `payment_sequential` | Payment method, installments and value |
| Reviews | One row per review record | `review_id` + `order_id` | Customer rating and written feedback |
| Products | One row per product | `product_id` | Product category and physical attributes |
| Sellers | One row per seller | `seller_id` | Seller identity and location |
| Geolocation | One row per geolocation record | No reliable single unique identifier identified | Geographic coordinates and location |
| Category Translation | One row per category translation | `product_category_name` | Portuguese-to-English category mapping |

## Important Analytical Fields

- `order_purchase_timestamp` → order timing
- `order_delivered_customer_date` → actual delivery
- `order_estimated_delivery_date` → estimated delivery
- `price` → product/item price
- `freight_value` → freight cost
- `review_score` → customer satisfaction
- `review_comment_message` → customer feedback
- `product_category_name` → product category
- `seller_id` → seller performance analysis
- `customer_unique_id` → repeat-customer analysis

# ⚠️ Phase 1 — Data Quality Observations

1. Several order date fields contain missing values, including approval and delivery dates.
2. 775 orders are not represented in Order Items.
3. One order is not represented in Payments.
4. 768 orders do not have a review record.
5. `review_id` contains 814 duplicate values, although the `review_id` + `order_id` combination is unique.
6. 610 products have missing `product_category_name`.
7. 13 product records belong to categories without English translations:
   - `portateis_cozinha_e_preparadores_de_alimentos` — 10 records
   - `pc_gamer` — 3 records
8. Geolocation contains 261,831 exact duplicate rows.
9. 278 customer records have ZIP prefixes not found in Geolocation.
10. 7 seller records have ZIP prefixes not found in Geolocation.
11. Order date columns are currently stored as `object` and require datetime conversion during data cleaning.
12. Estimated delivery dates extend beyond the latest observed actual customer delivery date.
13. Missing values and structural differences identified during Phase 1 will be investigated during Phase 2 rather than automatically removed.

### Phase 1 Principle

Observations are documented before any cleaning decisions are made. Missing values, duplicates, unmatched records and structural characteristics will be investigated based on their business meaning rather than automatically treated as errors.